internal/app/machined/pkg/controllers/network/address_status.go

```diff
@@ -0,0 +1,175 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/jsimonetti/rtnetlink"
+	"github.com/mdlayher/netlink"
+	"golang.org/x/sys/unix"
+	"inet.af/netaddr"
+
+	"github.com/talos-systems/talos/pkg/resources/network"
+	"github.com/talos-systems/talos/pkg/resources/network/nethelpers"
+)
+
+// AddressStatusController manages secrets.Etcd based on configuration.
+type AddressStatusController struct{}
+
+// Name implements controller.Controller interface.
+func (ctrl *AddressStatusController) Name() string {
+	return "network.AddressStatusController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *AddressStatusController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *AddressStatusController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: network.AddressStatusType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *AddressStatusController) Run(ctx context.Context, r controller.Runtime, logger *log.Logger) error {
+	watchConn, err := rtnetlink.Dial(&netlink.Config{
+		Groups: unix.RTMGRP_LINK | unix.RTMGRP_IPV4_IFADDR | unix.RTMGRP_IPV6_IFADDR,
+	})
+	if err != nil {
+		return fmt.Errorf("error dialing watch socket: %w", err)
+	}
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+
+	watchCh := make(chan struct{})
+
+	go func() {
+		defer wg.Done()
+
+		for {
```

Comment:
Should be have an internal deadline, or is the received context sufficiently time-bounded?